### PR TITLE
adds effective client request url var expansion

### PR DESCRIPTION
**Pull request by d2r**

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
[approve ci]

File: plugins/header_rewrite/expander.cc [hunk 1]
```diff
@@ -113,6 +113,13 @@ VariableExpander::expand(const Resources &res)
         }
         TSHandleMLocRelease(bufp, TS_NULL_MLOC, url_loc);
       }
+    } else if (variable == "%<cque>") {
+      // The client request effective URL.
+      int url_len = 0;
+      const char *url = TSHttpTxnEffectiveUrlStringGet(res.txnp, &url_len);
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/expander.cc]: This string is allocated and must be released. Bad API design, yes, but there it is.

Comment (d2r) [inline plugins/header_rewrite/expander.cc]: Yes, I did not notice that `TSHttpTxnEffectiveUrlStringGet` always calls `HTTPHdr::url_string_get` with a null first parameter. This causes new memory to be allocated for the resulting url string, and this must be freed. This is in contrast to calls to `TSHttpHdrMethodGet` and `TSUrlPathGet` in the two cases directly above in which the strings returned point to memory that is managed elsewhere.

**Comment by d2r** [comment]
Comment addressed.
Also added a clarifying comment for `TSHttpTxnEffectiveUrlStringGet` that callers must free.

File: plugins/header_rewrite/expander.cc (hunk 1 above)
Comment (SolidWallOfCode) [inline plugins/header_rewrite/expander.cc]: It's ugly, some day we'll have to clean that up, but backwards compatibility is an issue.

File: plugins/header_rewrite/expander.cc [hunk 2]
```diff
@@ -113,6 +113,15 @@ VariableExpander::expand(const Resources &res)
...
+      if (url && url_len) {
+        resolved_variable.assign(url, url_len);
+      }
+      free((void*)url);
```
Comment (SolidWallOfCode) [inline plugins/header_rewrite/expander.cc]: There's no need to cast here.

**Comment by d2r** [comment]
> There's no need to cast here.

The compiler complained of an illegal conversion of `const char *` to `char *`. As discussed, I removed the `const` and then squashed the commits.

**Comment by SolidWallOfCode** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci]

**Comment by d2r** [comment]
Fixed a clang-format error. I learned you can run `make clang-format` to discover which code style differences will trigger a failure in CI.

**Comment by persiaAziz** [comment]
[approve ci]

**Comment by persiaAziz** [comment]
[approve ci autest]

**Comment by d2r** [comment]
@SolidWallOfCode I think this is finally ready.

**Pull request merged**